### PR TITLE
feat: use jwt_access_token for new api (in development)

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

This is part of the work needed to allow the new api to a) authenticate users and b) accept requests from users that the old api authenticated.

I focused on two things in this PR

* creating fastify plugins to create and verify `jwt_access_token`s
* updating the mock login that's used in development and testing

The follow up PR(s) will

* use OpenID Connect to authenticate users (i.e. implement the old api's authentication flow in the new api)
* reduce the db usage by re-using `req.user` in the handlers, rather than fetching the user record twice
* update mobile-login to use `req.user` not `req.session.user`, since we don't have sessions.
* probably refactor a fair bit (names in particular)

Finally some design choices to bear in mind

* There's a small bug fix: the api handles the case where the user record is missing, but the `jwt_access_token` still exists (i.e. they were signed in on two devices and deleted their account)
* The cookie `jwt_access_token` and the access token encoded within it expire at different times. We should fix this post-MVP
* When redirecting the user I dropped query params i.e. if the referrer is `www.fcc.org/thing/?what=is`, it'll send you back to `www.fcc.org/thing/`. This seemed more sensible, since we _probably_ don't want to keep any old flash messages after redirecting, but it is a change.

### CODE REVIEW

File: api/src/utils/tokens.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * id of the UserToken document).
+ * @param userToken A token id to encode.
+ * @returns An encoded object with the userToken property.
+ */
+export function encodeUserToken(userToken: string): string {
+  return jwt.sign({ userToken }, JWT_SECRET);
+}
+
+// uppercase, lowercase letters and numbers
+const nanoid = customAlphabet(
```
Comment (moT01) [inline api/src/utils/tokens.ts:17]: We have this in the user file already. Could combine/move to neutral location?

**Review by moT01 — COMMENTED** [review]
Alright, I think this is looking good @ojeytonwilliams 🎉 

I looked over it pretty good. I left a number of comments with mostly minor things. I would potentially like to meet up sometime this week and just talk over it a bit if you have time.

File: api/src/routes/challenge.ts [hunk 2]
```diff
@@ -56,7 +56,7 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
   // @ts-expect-error - @fastify/csrf-protection needs to update their types
   // eslint-disable-next-line @typescript-eslint/unbound-method
   fastify.addHook('onRequest', fastify.csrfProtection);
-  fastify.addHook('onRequest', fastify.authenticateSession);
+  fastify.addHook('onRequest', fastify.authorize);
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/src/routes/challenge.ts:59]: Ah, good catch. Yes, they would be rejected.

I'll create an issue to track this.

File: api/config/allowed-origins.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export const allowedOrigins = [
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api/config/allowed-origins.ts:1]: I'll fix that separately.

**Comment by ojeytonwilliams** [comment]
Thanks for the review, @moT01. Working with the redirection helpers made me realise just how much they need to be updated, but I'll hold of until your PRs are in. Much as I want to fix it here!

Anyways, let me know if I missed anything.

**Review by moT01 — APPROVED** [review]
LGTM 🎉

File: api/src/routes/auth-dev.test.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    it('should populate the user with the correct data', async () => {
+      const uuidRe = /^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/;
+      const fccUuidRe = /^fcc-[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/;
+
+      await superRequest('/signin', { method: 'GET' });
+      const user = await fastifyTestInstance.prisma.user.findFirstOrThrow({
+        where: { email: defaultUserEmail }
+      });
+
+      expect(user).toMatchObject({
```
Comment (ShaunSHamilton) [inline api/src/routes/auth-dev.test.ts:48]: I realise this has just moved file. `toMatchObject` allows subsets to match. So, if we want to ensure all properties exist, we might want to use `toStrictEqual`.

**Review by ShaunSHamilton — APPROVED** [review]
Aside from a bug only in Gitpod (probably `https`/path related), this LGTM 👍

**Pull request merged**